test/service/test_tdr.py

```diff
@@ -0,0 +1,131 @@
+            self._make_mock_entity_table(dataset, 'links', [
+                dict(links_id=links_id, version=version, content='{}')
+                for version in versions
+                for links_id in links_ids
+            ])
+            client = self._init_client(dataset)
+            bundle_ids = client.list_links_ids(prefix='42')
+            bundle_ids.sort(key=attrgetter('uuid'))
+            self.assertEqual(bundle_ids, [
+                BundleFQID('42-abc', current_version),
+                BundleFQID('42-def', current_version),
+                BundleFQID('42-ghi', current_version)
+            ])
+
+        with self.subTest('snapshot'):
+            test(BigQueryDataset('test-project', 'name', True))
+        with self.subTest('dataset'):
+            test(BigQueryDataset('test-project', 'name', False))
+
+    def _make_mock_entity_table(self,
+                                dataset: BigQueryDataset,
+                                concrete_entity_type: str,
+                                rows: JSONs = (),
+                                additional_columns: Dict[str, Union[str, type]] = frozendict()):
+        self.query_adapter.create_table(dataset_name=dataset.name,
+                                        table_name=concrete_entity_type,
+                                        schema=self._bq_schema({f'{concrete_entity_type}_id': str},
+                                                               version='TIMESTAMP',
+                                                               content=str,
+                                                               **additional_columns),
+                                        rows=rows)
+
+    _bq_types = {
+        str: 'STRING',
+        bytes: 'BYTES',
+        int: 'INTEGER',
+        float: 'FLOAT',
+        bool: 'BOOL'
+    }
+
+    def _bq_schema(self, fields: Mapping[str, Union[str, type]] = frozendict(), **kwargs: Union[str, type]) -> JSONs:
+        kwargs.update(fields)
+        return [
+            dict(name=k,
+                 type=self._bq_types.get(v, v),
+                 mode='NULLABLE')
+            for k, v in kwargs.items()
+        ]
+
+
+class TinyBigQueryAdapter(AbstractBigQueryAdapter):
+
+    @cached_property
+    def client(self):
+        return tinyquery.TinyQuery()
+
+    def run_sql(self, query: str) -> Iterable[JSON]:
+        columns = self.client.evaluate_query(query).columns
+        num_rows = one(set(map(lambda c: len(c.values), columns.values())))
+        for i in range(num_rows):
+            yield {k[1]: v.values[i] for k, v in columns.items()}
+
+    def assert_table_exists(self, dataset_name: str, table_name: str) -> None:
+        self.client.get_table(dataset_name, table_name)
+
+    def create_table(self, dataset_name: str, table_name: str, schema: JSONs, rows: JSONs) -> None:
+        # TinyQuery's errors are typically not helpful in catching these mismatches
```

Comment:
What mismatches? Comment is not clear to me.

Reply:
Between row and table schemas. Updated.